### PR TITLE
Raise an early error on invalid labels in transitions inputs/outputs

### LINKED ISSUE #19632: Unrecoverable error using bzlmod
### Description of the bug:

While making rules_dotnet bzlmod compatible I encountered an Bazel crash:
```
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'ConfiguredTargetKey{label=//dotnet/private/tests/warning_settings:csharp_treat_warnings_as_errors_config_test, config=BuildConfigurationKey[0cfe0dbd956d9b571c331623c346c60b3b866c9f5bcd256de36fa22c2b46c2b8]}' (requested by nodes )
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:633)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:365)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.IllegalArgumentException: com.google.devtools.build.lib.cmdline.LabelSyntaxException: invalid repository name '@[unknown repo '' requested from @bazel_skylib~1.4.2]': repo names may contain only A-Z, a-z, 0-9, '-', '_', '.' and '~' and must not start with '~'
	at com.google.devtools.build.lib.cmdline.Label.parseAbsoluteUnchecked(Label.java:240)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(Unknown Source)
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline.collect(Unknown Source)
	at com.google.devtools.build.lib.analysis.starlark.StarlarkTransition.getRelevantStarlarkSettingsFromTransition(StarlarkTransition.java:393)
	at com.google.devtools.build.lib.analysis.starlark.StarlarkTransition.lambda$getAllStarlarkBuildSettings$0(StarlarkTransition.java:123)
	at com.google.devtools.build.lib.analysis.starlark.StarlarkTransition$StarlarkTransitionVisitor.accept(StarlarkTransition.java:430)
	at com.google.devtools.build.lib.analysis.config.transitions.ConfigurationTransition.visit(ConfigurationTransition.java:101)
	at com.google.devtools.build.lib.analysis.config.transitions.ComposingTransition.visit(ComposingTransition.java:99)
	at com.google.devtools.build.lib.analysis.config.transitions.ComposingTransition.visit(ComposingTransition.java:99)
	at com.google.devtools.build.lib.analysis.starlark.StarlarkTransition.getAllStarlarkBuildSettings(StarlarkTransition.java:119)
	at com.google.devtools.build.lib.analysis.config.ConfigurationResolver.getStarlarkBuildSettingsDetailsValue(ConfigurationResolver.java:407)
	at com.google.devtools.build.lib.analysis.config.ConfigurationResolver.applyTransitionWithSkyframe(ConfigurationResolver.java:388)
	at com.google.devtools.build.lib.analysis.config.ConfigurationResolver.resolveGenericTransition(ConfigurationResolver.java:233)
	at com.google.devtools.build.lib.analysis.config.ConfigurationResolver.resolveConfiguration(ConfigurationResolver.java:192)
	at com.google.devtools.build.lib.analysis.config.ConfigurationResolver.resolveConfigurations(ConfigurationResolver.java:151)
	at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.computeDependencies(ConfiguredTargetFunction.java:869)
	at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.compute(ConfiguredTargetFunction.java:371)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:562)
	... 7 more
Caused by: com.google.devtools.build.lib.cmdline.LabelSyntaxException: invalid repository name '@[unknown repo '' requested from @bazel_skylib~1.4.2]': repo names may contain only A-Z, a-z, 0-9, '-', '_', '.' and '~' and must not start with '~'
	at com.google.devtools.build.lib.cmdline.LabelParser.syntaxErrorf(LabelParser.java:208)
	at com.google.devtools.build.lib.cmdline.RepositoryName.validate(RepositoryName.java:160)
	at com.google.devtools.build.lib.cmdline.LabelParser$Parts.validateRepoName(LabelParser.java:180)
	at com.google.devtools.build.lib.cmdline.LabelParser$Parts.validateAndCreate(LabelParser.java:72)
	at com.google.devtools.build.lib.cmdline.LabelParser$Parts.parse(LabelParser.java:164)
	at com.google.devtools.build.lib.cmdline.Label.parseCanonical(Label.java:123)
	at com.google.devtools.build.lib.cmdline.Label.parseAbsoluteUnchecked(Label.java:238)
	... 30 more
```

Reproduction is available at https://github.com/bazelbuild/rules_dotnet/pull/387 by running `bazel test //... --enable_bzlmod`

cc @fmeum

### Which category does this issue belong to?

_No response_

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

_No response_

### Which operating system are you running Bazel on?

Linux Fedora 38

### What is the output of `bazel info release`?

release 6.3.2

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

_No response_

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

_No response_

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.

_No response_

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

_No response_

### PR DESCRIPTION
**Pull request by fmeum**
When a label pointing to an unavailable repository is specified as a transition's input or output, an early error is raised instead of crashing during transition evaluation.

Fixes #19632

### CODE REVIEW

File: src/test/java/com/google/devtools/build/lib/analysis/RunfilesRepoMappingManifestTest.java [hunk 1]
```diff
@@ -435,4 +435,45 @@ public void overwriteWorkspaceFile(String... lines) throws Exception {
             .addAll(ImmutableList.copyOf(lines))
             .build());
   }
+
+  @Test
+  public void testInvisibleRepoInLabelResultsInEarlyError() throws Exception {
```
Comment (fmeum) [inline src/test/java/com/google/devtools/build/lib/analysis/RunfilesRepoMappingManifestTest.java:440]: I couldn't find a generic Bzlmod `BuildViewTestCase` test. Should I just create a new one?

**Comment by fmeum** [comment]
@bazel-io flag

**Comment by iancha1992** [comment]
@bazel-io fork 6.4.0

File: src/test/java/com/google/devtools/build/lib/analysis/RunfilesRepoMappingManifestTest.java (hunk 1 above)
Comment (Wyverald) [reply src/test/java/com/google/devtools/build/lib/analysis/RunfilesRepoMappingManifestTest.java:440]: https://cs.opensource.google/bazel/bazel/+/master:src/test/java/com/google/devtools/build/lib/rules/LabelBuildSettingTest.java this is where I parked related tests

File: src/main/java/com/google/devtools/build/lib/rules/config/ConfigGlobalLibrary.java [hunk 2]
```diff
@@ -97,8 +106,11 @@ private void validateBuildSettingKeys(
     for (String optionKey : optionKeys) {
       if (!optionKey.startsWith(COMMAND_LINE_OPTION_PREFIX)) {
         try {
-          var unused = Label.parseCanonicalUnchecked(optionKey);
-        } catch (IllegalArgumentException e) {
+          Label label = Label.parseWithRepoContext(optionKey, packageContext);
+          if (!label.getRepository().isVisible()) {
+            throw Starlark.errorf("invalid transition %s '%s'", singularErrorDescriptor, label);
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/rules/config/ConfigGlobalLibrary.java:111]: maybe add the whole clause about "repo X not visible from X"?

**Review by Wyverald — APPROVED** [review]
please import after comments are addressed!

File: src/main/java/com/google/devtools/build/lib/rules/config/ConfigGlobalLibrary.java (hunk 2 above)
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/rules/config/ConfigGlobalLibrary.java:111]: Done

**Comment by fmeum** [comment]
@iancha1992 Comments resolved

**Pull request closed without merging**